sklearn/utils/_metadata_requests.py

```diff
@@ -0,0 +1,1091 @@
+"""
+Metadata Routing Utility
+"""
+
+# Author: Adrin Jalali <adrin.jalali@gmail.com>
+# License: BSD 3 clause
+
+import inspect
+from copy import deepcopy
+from enum import Enum, EnumMeta
+from warnings import warn
+from collections import namedtuple
+from typing import Union, Optional
+from ._bunch import Bunch
+
+# This namedtuple is used to store a (mapping, routing) pair. Mapping is a
+# MethodMapping object, and routing is the output of `get_metadata_routing`.
+# MetadataRouter stores a collection of these namedtuples.
+RouterMappingPair = namedtuple("RouterMappingPair", ["mapping", "router"])
+
+# A namedtuple storing a single method route. A collection of these namedtuples
+# is stored in a MetadataRouter.
+MethodPair = namedtuple("MethodPair", ["method", "used_in"])
```

Comment:
Looked through `doc/metadata_routing.rst` again. We do use the term "consumer" for anything that accepts metadata.

- "used_in" -> "consumer" or "receiver"
- "method" -> "producer" or "provider"

In this mental model, the "provider" method provides the metadata and "receiver" method receives the metadata.

Reply:
"consumer" is distinct from ?"receiver". The consumer needs to know how to use the passed prop, and is able to have a request set on it. A receiver that only routes does not need either of these properties.

I also think you've got things backwards....? I'm confused about the name "provider" or "producer".

The MethodPair is intended to describe how a metaestimator delegates functionality in its `used_in` method to a wrapped estimator by calling its `method`. So `used_in` more-or-less means "caller", and `method` means "callee". The callee passes its request to the estimator containing the caller. The caller, given the metadata, provides it to the callee. You could call these "requestor" and "provider", or "base_method" and "meta_method", or "wrapped" and "wrapper".

Reply:
> I also think you've got things backwards....?

Yea I got it backwards. I confused the semantics of `method` and `uesd_in` again.

I'm happy with "caller" and "callee".

Reply:
Requestor and provider are more the object themselves rather than the methods in my mind.

Reply:
Also, summarizing some of the points from the meeting @thomasjpfan, @glemaitre and I had a few days ago:

- `method, used_in` can be confusing since one can think of the `used_in` as "the method used in the other one", or "the data used in the other one".
- `src, dst` can be confusing since one can think of the data itself being routed, or the request being sent from the child to the parent, which makes it confusing. They're also terms that people with computer science background will understand much easier than others.

We settled at the end on `parent_method, child_method`, at at the end of the meeting we were somewhat happy with them as a compromise, and they seemed quite clear.

Out of the new options mentioned here, I'd be happy with `caller, callee` the most. Look at the following:

``` python
        router = (
            MetadataRouter()
            .add(
                transformer=self.transformer,
                method_mapping=MethodMapping()
                .add(callee="fit", caller="fit")
                .add(callee=transform", caller="fit")
                .add(callee="transform", caller="predict"),
            )
            .add(classifier=self.classifier, method_mapping="one-to-one")
        )
```


and if you compare that with:

``` python
        router = (
            MetadataRouter()
            .add(
                transformer=self.transformer,
                method_mapping=MethodMapping()
                .add(requestor="fit", provider="fit")
                .add(requestor="transform", provider="fit")
                .add(requestor="transform", provider="predict"),
            )
            .add(classifier=self.classifier, method_mapping="one-to-one")
        )
```

the latter has confusing semantics, as `transform` is requesting twice, and two different methods are now providing. What's happening in the implementation is more of who's calling who rather than who provides what I think. To some extend same goes to the other options.

Reply:
> @thomasjpfan you changed your mind from caller, callee to requestor and provider, why?

I thought the words looked too similar to each other. Looking at your code example, I think it's fine. Let's go with `callee` and `caller`.

Reply:
Closing this thread for now, we seem to have kind of a consensus on this.